### PR TITLE
[wasm-split] Add a multi-split mode

### PR DESCRIPTION
**Pull request by tlively**
Add a mode that splits a module into arbitrarily many parts based on a
simple manifest file. This is currently implemented by splitting out one
module at a time in a loop, but this could change in the future if
splitting out all the modules at once would improve the quality of the
output.

### CODE REVIEW

**Comment by tlively** [comment]
cc @dschuff FYI

File: src/tools/wasm-split/wasm-split.cpp [hunk 1]
```diff
@@ -362,6 +362,82 @@ void splitModule(const WasmSplitOptions& options) {
...
+
+  std::ifstream manifest(options.manifestFile);
+  if (!manifest.is_open()) {
+    Fatal() << "File not found: " << options.manifestFile;
+  }
+
+  Module wasm;
+  parseInput(wasm, options);
+
+  std::map<std::string, std::unordered_set<std::string>> moduleFuncs;
```
Comment (kripken) [inline src/tools/wasm-split/wasm-split.cpp:381]: Not hard to figure these out, but some comments would be nice I think.

File: src/tools/wasm-split/wasm-split.cpp [hunk 2]
```diff
@@ -362,6 +362,82 @@ void splitModule(const WasmSplitOptions& options) {
...
+    }
+    if (!options.quiet && funcs.empty()) {
+      std::cerr << "warning: Module " << mod << " will be empty\n";
+    }
+    for (auto& func : funcs) {
+      config.primaryFuncs.erase(Name(func));
+    }
+    auto splitResults = ModuleSplitting::splitFunctions(wasm, config);
+    // TODO: symbolMap, placeholderMap, emitModuleNames
+    // TODO: Support --emit-text and use .wast in that case.
```
Comment (kripken) [inline src/tools/wasm-split/wasm-split.cpp:431]: Maybe error on these for now?

File: src/tools/wasm-split/wasm-split.cpp [hunk 3]
```diff
@@ -362,6 +362,82 @@ void splitModule(const WasmSplitOptions& options) {
...
+    auto moduleName = options.outPrefix + mod + ".wasm";
+    PassRunner runner(&*splitResults.secondary);
+    runner.add("remove-unused-module-elements");
```
Comment (kripken) [inline src/tools/wasm-split/wasm-split.cpp:434]: Why is this needed here but not in single-split?

Comment (tlively) [reply src/tools/wasm-split/wasm-split.cpp:434]: We should probably add it to the normal split mode as well. In experiments with hundreds of relatively small modules, this was obviously very necessary. I guess it's harder to tell by visual inspection that it is necessary for larger secondary modules.

File: src/tools/wasm-split/wasm-split.cpp (hunk 2 above)
Comment (tlively) [reply src/tools/wasm-split/wasm-split.cpp:431]: Our special option validation that checks that each flag is allowed in the selected mode already takes care of that.

File: src/tools/wasm-split/wasm-split.cpp (hunk 3 above)
Comment (kripken) [reply src/tools/wasm-split/wasm-split.cpp:439]: Makes sense. Maybe add a TODO for the normal mode?

Comment (tlively) [reply src/tools/wasm-split/wasm-split.cpp:439]: Just uploaded #6945 to do this for all secondary modules.

**Review by kripken — APPROVED** [review]
